### DISCUSSION TITLE
Awaiting OptionAsync / EitherAsync

### DISCUSSION THREAD

**Discussion by BrunoJuchli** [louthy/language-ext#960] [Q&A]
## EitherAsync
In contrast to `OptionAsync<A>`, `EitherAsync<L,R>` can't be await because it's missing the `AsyncMethodBuilderAttribute` -- well it's actually commented out as can be seen here:
https://github.com/louthy/language-ext/blob/d5248b667254548e32051db59b5ca2d6672389e6/LanguageExt.Core/Monads/Alternative%20Value%20Monads/Either/EitherAsync/EitherAsync.cs#L35

Is there a plan to add that later on? Are there significant obstacles to doing so or would a PR be accepted?
(I'd expect awaiting an `EitherAsync<L,R>` to return an `Either<L,R>`).

## OptionAsync
Why does awaiting an `OptionAsync<A>` return `A?` and not `Option<A>`?

The questions arose in the context of a unit test, like:

     var result = await testee.DoSomething(); // DoSomething has a signature return-type of OptionAsync<A>()
     
     result.Should().BeNone(); // doesn't work because it's null

Now for the test it doesn't matter that much, I can usually just test for `null` instead  -- i still don't like it very much because IMO it's confusing to work with `null` when explicitly trying to go the extra mile to work funtional style.
So what I do is I add a `.ToOption()` after the call to `DoSomething()`. I'd rather expected it to be vice-versa: requiring an extra call like "ToNullable()" to have it return a `Task<A?>`.
Maybe there's a good reason for the current behavior and someone can shed some light on it? Always happy to learn :-)

**Comment by fjmorel** [comment]
I forgot to check discussions before writing up https://github.com/louthy/language-ext/issues/1019

You _can await_ an `EitherAsync`, but you can't use it as the return type of an async method yet, because of the commented out line you pointed out.

edit: copying link to an issue which explains the async/await limitation: https://github.com/dotnet/csharplang/issues/3723
